File: pset6/credit/credit.py
from cs50 import get_string

while True:
    card = get_string("Please insert credit card number: ")
    if card.isdigit() == False:
        continue
    number = int(card)
    if number > 0:
        break

lenght = len(card)

# Size-based validation
if lenght != 13:
    if lenght != 15:
        if lenght != 16:
            result = "INVALID"

cnumbers = []
sumn = 0

for i in card:
    cnumbers.append(i)

# Algorithm-based validation
# 1st step for validation: Multiplication by 2
rest = cnumbers[::-2]

for i in range(0, lenght - 1, 2):
    if (int(cnumbers[lenght - 2 - i]) * 2) > 9:
        sumn += (int(cnumbers[lenght - 2 - i]) * 2) - 9
    else:
        sumn += (int(cnumbers[lenght - 2 - i]) * 2)

# 2nd step for validation: sums all digits
for i in rest:
    sumn += int(i)
print(sumn)

# 3rd step for validation: checksum with final 0
if (sumn % 10) != 0:
    result = "INVALID"
elif card[0] == "4":
    result = ("VISA")
elif card[0] == "5":
    if int(card[1]) > 0 and int(card[1]) < 6:
        result = ("MASTERCARD")
elif card[0] == "3":
    if card[1] == "4" or card[1] == "7":
        result = ("AMEX")
    else:
        result = ("INVALID")
else:
    result = ("INVALID")

print(result)

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

int main()
{
    int card[16];
    int digits = 0, size = 0, sum = 0;
    long int number;
    
    do
    {
        number = get_long("Please insert credit card number: ");
    }
    while (number < 0);
    
    //This block determines the lenght of credit card number
    long int counter = number;
    while (counter != 0)
    {
        counter = counter / 10;
        size++;
    }
    switch (size)
    {
        case 13:
            digits = 3;
            break;
            
        case 15:
            digits = 1;
            break;
            
        case 16:
            digits = 0;
            break;
            
        default:
            printf("ERROR! Non-valid number.");
    } printf("\n\nsize: %i and digits: %i\n\n", size, digits);
    
    for (int i = (15 - digits); i >= 0; i--)
    {
        card[i] = number % 10;
        number = number / 10;
    }

    printf("\n");
    
    //1st step for validation: Multiplication by 2
    for (int i = (14 - digits); i >= 0; i -= 2)
    {
        sum = sum + (card[i] * 2 / 10) + ((card[i] * 2) % 10);
    }
    printf("1st step validation: %i\n\n", sum);
    
    //2nd step for validation: checksum with final 0
    for (int i = (15 - digits); i >= 0; i -= 2)
    {
        sum = sum + card[i];
    }
    printf("2nd step validation: %i\n\n", sum);
    
    if ((sum % 10) != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        if (card[0] == 4)
        {
            printf("VISA\n");
        }
        else if (card[0] == 5)
        {
            if (card[1] == 1 || card[1] == 2 || card[1] == 3 || card[1] == 4 || card[1] == 5)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (card[0] == 3)
        {
            if (card[1] == 4 || card[1] == 7)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}
